Wikipedia/Code/ArticlesCollectionViewController.swift

```diff
@@ -0,0 +1,482 @@
+    let dataStore: MWKDataStore
+    var fetchedResultsController: NSFetchedResultsController<ReadingListEntry>?
+    var collectionViewUpdater: CollectionViewUpdater<ReadingListEntry>?
+    let readingList: ReadingList
+    
+    var searchString: String?
+    
+    var basePredicate: NSPredicate {
+        return NSPredicate(format: "list == %@ && isDeletedLocally != YES", readingList)
+    }
+    
+    var shouldShowEditButtonsForEmptyState: Bool {
+        return !readingList.isDefault
+    }
+    
+    var searchPredicate: NSPredicate? {
+        guard let searchString = searchString else {
+            return nil
+        }
+        return NSPredicate(format: "(displayTitle CONTAINS[cd] '\(searchString)')")
+    }
+    
+    var availableBatchEditToolbarActions: [BatchEditToolbarAction] {
+        return [
+            BatchEditToolbarActionType.addTo.action(with: nil),
+            BatchEditToolbarActionType.moveTo.action(with: nil),
+            BatchEditToolbarActionType.remove.action(with: nil)
+        ]
+    }
+    
+    var editController: CollectionViewEditController!
+    
+    private var cellLayoutEstimate: ColumnarCollectionViewLayoutHeightEstimate?
+    private let reuseIdentifier = "ArticlesCollectionViewCell"
+    
+    weak var delegate: ArticlesCollectionViewControllerDelegate?
+    
+    init(for readingList: ReadingList, with dataStore: MWKDataStore) {
+        self.readingList = readingList
+        self.dataStore = dataStore
+        super.init()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        layoutManager.register(SavedArticlesCollectionViewCell.self, forCellWithReuseIdentifier: reuseIdentifier, addPlaceholder: true)
+        setupEditController()
+        isRefreshControlEnabled = true
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        setupFetchedResultsController()
+        setupCollectionViewUpdater()
+        fetch()
+        super.viewWillAppear(animated)
+    }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        collectionViewUpdater = nil
+        fetchedResultsController = nil
+        editController.close()
+    }
+    
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        cellLayoutEstimate = nil
+    }
+    
+    func configure(cell: SavedArticlesCollectionViewCell, for article: WMFArticle, at indexPath: IndexPath, layoutOnly: Bool) {
+        cell.isBatchEditing = editController.isBatchEditing
+        
+        guard let entry = readingList.entry(for: article) else {
```

Comment:
For reading lists with a lot of entries, this search over the entire list of entries to get the entry for a given article is probably too slow to call for every cell configuration.

This function could take a `ReadingListEntry`, because it can be looked up quickly from the `fetchedResultsController`. Then the article can be looked up using `dataStore.fetchArticle(` (which is cached) as it is in `article(at:`